application/apps/indexer/sources/src/serial/serialport.rs

```diff
@@ -0,0 +1,106 @@
+use crate::{Error as SourceError, ReloadInfo, SourceFilter};
+use crate::ByteSource;
+use buf_redux::Buffer;
+use tokio_stream::StreamExt;
+use std::time::Duration;
+use tokio_serial::{SerialStream, SerialPortBuilderExt};
+use tokio::io::AsyncReadExt;
+use async_trait::async_trait;
+
+const TIMEOUT: u64 = 100;
+const MINIMAL_PACKAGE_SIZE: usize = 255;
+
+pub struct SerialSource {
+    port: SerialStream,
+    buffer: Buffer,
+    read: [u8; 1],
+    amount: usize,
+}
+
+impl SerialSource {
+    pub fn new(path: &str, baud_rate: u32) -> Self {
+        Self {
+            buffer: Buffer::new(),
+            read: [0; 1],
+            amount: 0,
+            port: tokio_serial::new(path, baud_rate).open_native_async()
+                .expect("open port failed"),
+        }
+    }
+}
+
+#[async_trait]
+impl ByteSource for SerialSource {
+    async fn reload(&mut self, _filter: Option<&SourceFilter>) ->  Result<Option<ReloadInfo>, SourceError> {
+        let mut copied: usize = 0;
+        let _ = tokio::time::timeout(Duration::from_millis(TIMEOUT), async {
+            loop {
+                self.port.read_exact(&mut self.read).await.expect("read from serial stream failed");
+                copied += self.buffer.copy_from_slice(&self.read);
+                if copied > MINIMAL_PACKAGE_SIZE {
+                    break;
+                }
+            }
+        }).await;
+        self.amount = copied;
+        Ok(Some(ReloadInfo::new(
+            self.amount,
+            self.buffer.free_space(),
```

Comment:
this part of the reload info should contain the amount of bytes that could currently be used by the `MessageProducer`...
so in your case the amount of bytes that were received and not consumed so far.
The `free_space()` will return the total amount of free space in the buffer, including bytes already consumed from the head.